### PR TITLE
Proxy sync fewer endpoints

### PR DESCRIPTION
**Pull request by dcbw**
This is a follow-on to https://github.com/kubernetes/kubernetes/pull/38996 but the two are largely independent.  As such, it includes those commits too but they can be disregarded and I'll rebase if the other one changes.

Build the endpoints map in a separate testable function. Return that map instead of changing proxier.endpointsMap directly. Use reflect.DeepEqual() to skip syncing proxy rules if nothing actually changed.

@thockin @kubernetes/rh-networking @kubernetes/sig-network-misc @timothysc @wojtek-t @jeremyeder @caseydavenport

### CODE REVIEW

File: pkg/proxy/iptables/proxier.go
Comment (knobunc) [inline pkg/proxy/iptables/proxier.go]: It looks like the code in DeepEqual (once it determines it is a map in the helper function it calls) tests the length fairly early on in the function.

Is there an advantage to having this function rather than just calling DeepEqual directly?

**Review by knobunc — CHANGES_REQUESTED** [review]
LGTM, I just have a minor question.  But I like the approach.  Thanks!

File: pkg/proxy/iptables/proxier.go
Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: @knobunc added per @thockin request in https://github.com/kubernetes/kubernetes/pull/38996#pullrequestreview-14105729.  I guess it's not computationally simple to determine each variables type in reflect or something?

Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: @knobunc and he seems to be right; for a simple program comparing two map[string]string with 6 keys in one and 8 keys in another, the difference between using len() and reflect.DeepEqual over 5000 iterations is:

DeepEqual: 522.389µs
len: 4.904µs

**Comment by knobunc** [comment]
Wow... thanks for that speed comparison.  That's an impressive savings.

**Comment by thockin** [comment]
I am confused on this - you have commits from the other PR, but there's something wonky - it looks like those commits are in here twice, or you used the EXACT SAME names for 4 commits.  Can you sanity check it and help me figure what to review here?

**Comment by danwinship** [comment]
> it looks like those commits are in here twice, or you used the EXACT SAME names for 4 commits

They're not exactly the same: "service" vs "endpoints". The new commits start with "proxy/iptables: split out endpoint map creation and add testcases"

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: Heh. I didn't mean to literally use that text, although I suppose it's true...
(s/container/containing/)

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: (following up from comments on previous PR). OK, I see, you're distinguishing `endpointsMap[port.Name]` being implicitly `nil` vs being explicitly set to `nil`. Maybe the comment could clarify that better (the fact that it will affect an iteration over the map later). Also, line width is way too short. (Unless you were aiming for a haiku... "Is address valid? / Remember invalid ports / To clean up later.")

**Comment by dcbw** [comment]
> I am confused on this - you have commits from the other PR, but there's something wonky - it looks like those commits are in here twice, or you used the EXACT SAME names for 4 commits. Can you sanity check it and help me figure what to review here?

@thockin as it said in the first comment on the, this one originally included all the commits from the "sync fewer services" PR, becuase I cannot figure out how to get github to submit a PR against kubernetes that itself is based off a different PR.  Happy to be educated.

But it's moot now that the services PR has been merged, so I have rebased this PR onto git master and those service sync commits are now gone.

**Comment by dcbw** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dcbw** [comment]
@k8s-bot gci gke e2e test this

**Comment by dcbw** [comment]
@thockin thoughts on the latest iteration now that it's rebased on top of the service changes?

**Comment by dcbw** [comment]
@thockin rebased to resolve conflict, PTAL thanks!

**Comment by dcbw** [comment]
@thockin rebased on top of @freehan changes, which drops one of the commits.

**Comment by dcbw** [comment]
cc @bprashanth

**Comment by thockin** [comment]
This is still huge - is there any reduction possible before jumping into review ?  :)

File: pkg/proxy/iptables/proxier.go
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: I know it's not required by language and it makes things  wee bit more syntactically heavy, but I prefer passing maps that we expect the callee to modify by pointer, so it's clearer at the callsite

**Review by thockin — COMMENTED** [review]
I started reviewing this and then I realized I need to let Minhan do it.  I am too buried right now.

HOWEVER, this is a big conflict with #41022, and I am trying to figure out whose is easier to rebase on the other.  I *think* this PR is easier to rebase on top of #41022 (once it is done being rebased to head).

Thoughts?

**Comment by dcbw** [comment]
@thockin allow me a short rant... this PR has been open for over a month, and yet there are multiple PRs that touch this same code and attempt to do very very similar things.  What can we do to reduce duplication of effort in this area and save all of us from continuous, painful rebases?

**Comment by timothysc** [comment]
> allow me a short rant...  this PR has been open for over a month

This is a pretty serious issue that needs to get resolved at the project level, and I think hits at the heart of some of the reasons @brendandburns has created his contributor health survey.  I think @thockin should be able to go on vacation and spend time with his family and the ship keep moving.  I almost believe that we need a project wide level of redundancy so that we don't ask for super-human feats from a few core contributors.  

/cc @kubernetes/sig-contributor-experience-misc

**Comment by bprashanth** [comment]
You just need to convince one of the OWNERS to LGTM it, then own it yourself by paying attention to the e2e flakes for related Services tests that can block SQ pre and post merge for a week or so. @freehan probably has the most context in this case.

**Comment by freehan** [comment]
Just scan thru a few PRs opened against kubeproxy.

This PR has 90% overlap with #41022 and shared some pieces with #40868. 

How about this? 
Since #41022 went further and has more test, we let it in first. Then apply this PR on top of it. IIUC, at that point, the main change will be refactoring the structures and healthcheck stuff.

**Comment by thockin** [comment]
Regarding the overlap - I think we all came at different facets of the
problem, but they happen to touch the same code.  This sucks.  I suck.  I
am just overwhelmed trying to juggle, and this whole family of PRs got
hard...

**Comment by feiskyer** [comment]
> How about this?
Since #41022 went further and has more test, we let it in first. Then apply this PR on top of it. IIUC, at that point, the main change will be refactoring the structs and healthcheck stuff.

Agreed with @freehan, since there are already overlapped PRs, it's better to make one in to unblock others.

**Comment by timothysc** [comment]
> This sucks.  I suck.  I am just overwhelmed trying to juggle, and this whole family of PRs got hard...

@thockin nonsense, you are indispensable... We just need to figure out how to either clone you, or delegate to the sig-leads.

**Comment by thockin** [comment]
Yeah, I'm working on delegation.

kube-proxy has been "my baby" so long, I wrestle with this one.  I thought
I was helping by taking up your PR and carrying the idea forward, but I
made something of a mess for poor Dan.

We do need a couple people who feel they have time and willingness to "own"
kube-proxy (and then we need to have some deep discussions about roadmap)
so I can be less involved in it.

On Thu, Feb 16, 2017 at 6:21 AM, Timothy St. Clair <notifications@github.com

**Comment by dcbw** [comment]
@thockin @freehan @bprashanth I'm happy to rework a PR and change approaches based on advice, which means other people have less work to do.  What sucks is when nobody looks at it for *5 weeks*, and then they write other PRs that hugely overlap with this one, and then theirs go in first and I have to do major rebases.  I like contributing to Kubernetes, but this makes it less fun.

Might it not be less work for everyone to do reviews and suggest changes, and let somebody else (like me!) do the work?

**Comment by dcbw** [comment]
I don't think this PR has anything to offer now that https://github.com/kubernetes/kubernetes/pull/41022 has been merged.  It would have been pretty great to know that 41022 was going to solve this problem 5 weeks ago, so I could have spent my time on something else, but whatever.

**Comment by thockin** [comment]
41022 wasn't intended to exist - I was tracking Tim's async-run work and
saw a problem, which required a bit of this, and then I didn't realize the
overlap.

I honestly do feel really bad, and I owe you a beer or something.  I just
saw your good feedback, and I will get on it.

**Pull request closed without merging**
